File: frontend/types/budget.ts
export interface CategorySpending {
  category: string;
  budget: number;
  spent: number;
  remaining: number;
  percentage: number;
}

export interface Budget {
  id: string;
  user_id: string;
  month: number;
  year: number;
  monthly_limit: string;
  category_limits: Record<string, number>;
  rollover_enabled: boolean;
  savings_goal: string;
  created_at: string;
  updated_at: string;
}

export interface BudgetWithSpending extends Budget {
  spending_so_far: string;
  remaining_budget: string;
  percentage_used: number;
  daily_average: number;
  days_elapsed: number;
  days_in_month: number;
  category_spending: CategorySpending[];
}

export interface BudgetHistoryItem extends Budget {
  actual_spending: number;
  percentage_used: number;
}

export interface BudgetCreate {
  monthly_limit: number;
  category_limits?: Record<string, number>;
  savings_goal?: number;
  rollover_enabled?: boolean;
}

export interface BudgetUpdate {
  monthly_limit?: number;
  category_limits?: Record<string, number>;
  savings_goal?: number;
  rollover_enabled?: boolean;
}

File: backend/app/schemas/budget.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, ConfigDict, field_validator


class BudgetCreate(BaseModel):
    monthly_limit: Decimal
    category_limits: dict[str, float] = {}
    savings_goal: Optional[Decimal] = Decimal("0")
    rollover_enabled: bool = False

    @field_validator("monthly_limit")
    @classmethod
    def validate_limit(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("Monthly limit must be greater than 0")
        return v


class BudgetUpdate(BaseModel):
    monthly_limit: Optional[Decimal] = None
    category_limits: Optional[dict[str, float]] = None
    savings_goal: Optional[Decimal] = None
    rollover_enabled: Optional[bool] = None


class BudgetSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    month: int
    year: int
    monthly_limit: Decimal
    category_limits: dict
    rollover_enabled: bool
    savings_goal: Decimal
    created_at: datetime
    updated_at: datetime


class CategorySpendingSchema(BaseModel):
    category: str
    budget: float
    spent: float
    remaining: float
    percentage: float


class BudgetWithSpendingSchema(BudgetSchema):
    spending_so_far: Decimal
    remaining_budget: Decimal
    percentage_used: float
    daily_average: float
    days_elapsed: int
    days_in_month: int
    category_spending: list[CategorySpendingSchema]


class BudgetHistoryItem(BudgetSchema):
    actual_spending: float
    percentage_used: float
